### PR TITLE
 [Idea]: SMAWK Algorithm implementation #1515 

### LINKED ISSUE #1515: [Idea]: SMAWK Algorithm implementation
### Idea Title

SMAWK Algorithm (Dynamic Programming Optimization) implementation 

### Idea Description

Add this algo to have more knowledge uncover

### Potential Benefits

- [X] Improves user experience
- [X] Enhances learning
- [X] Adds a new feature
- [ ] Improves code quality
- [X] Other

### Implementation Suggestions (Optional)

Assign me this under hacktoberfest and gssoc-ext.

### PR DESCRIPTION
**Pull request by AE-Hertz**
### Description
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

 [Idea]: SMAWK Algorithm implementation #1515 

Fixes #1515

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/algorithms/SMAWK Algorithm/SMAWK Algorithm.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
+---
+
+id: SMAWK-Algorithm
```
Comment (ajay-dhangar) [inline docs/algorithms/SMAWK Algorithm/SMAWK Algorithm.md]: replace `id: SMAWK-Algorithm` with `id: smawk-algorithm`

File: docs/algorithms/SMAWK Algorithm/SMAWK Algorithm.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+- **Algorithm Type**: Dynamic Programming Optimization, Matrix Algorithms.
+- **Main Operation**: Finds the minimum element in each row of a totally monotone matrix.
+- **Data Structures**: A vector to store the minimum values and another to track the active set of columns.
+- **Output**: A vector containing the minimum value of each row in the matrix.
+
+### Time Complexity
+
+- **Average Case**: \( O(n \log n) \), where \( n \) is the number of rows in the matrix.
+- **Worst Case**: \( O(n m) \), where \( m \) is the number of columns in the matrix.
+
```
Comment (ajay-dhangar) [inline docs/algorithms/SMAWK Algorithm/SMAWK Algorithm.md]: replace `\( O(n \log n) \)` with `$O(n \log n) $`, `\( n \)`  with `$n$` , `\( O(n m) \)` with `$O(m n)$`, and `\( m \)` with `$m$`

File: docs/algorithms/SMAWK Algorithm/SMAWK Algorithm.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+### Space Complexity
+
+- **Space Complexity**: \( O(n) \), since only a few vectors (of size \( n \)) are used to store the row minima and the active set.
+
```
Comment (ajay-dhangar) [inline docs/algorithms/SMAWK Algorithm/SMAWK Algorithm.md]: use `$`

File: docs/algorithms/SMAWK Algorithm/SMAWK Algorithm.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+### Approach
+
+1. **Input**: The algorithm accepts a matrix of size \( n \times m \).
```
Comment (ajay-dhangar) [inline docs/algorithms/SMAWK Algorithm/SMAWK Algorithm.md:33]: use `$`

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update your code based on suggestions

**Comment by AE-Hertz** [comment]
> update your code based on suggestions

resolved

**Review by ajay-dhangar — APPROVED** [review]
Done

**Pull request merged**